### PR TITLE
Update README.md with instructions on integrating Mechs

### PR DESCRIPTION
**Pull request by adrianoVal**
Updates the readme with instructions on integrating Mechs into you application and the mech request flow diagram

### CODE REVIEW

**Review by dvilelaf — COMMENTED** [review]
LGTM. Is the intro part going to live on the website?

**Comment by adrianoVal** [comment]
> LGTM. Is the intro part going to live on the website?

It will eventually, first we'll just link to the information from the site

**Review by dvilelaf — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -38,6 +38,51 @@ The project consists of three components:
   - Gives an overview of the AI workers in the registry.
   - Allows Mech owners to create new workers.
   - Allows users to request work from an existing worker.
+ 
+## Mech request-response flow
+
+![image](https://github.com/adrianoVal/mech/assets/90184519/c93a0096-b102-4414-822f-567e1fe719f5)
```
Comment (jmoreira-valory) [inline README.md:44]: Please, store images in the /docs/images folder

File: README.md [hunk 2]
```diff
@@ -38,6 +38,51 @@ The project consists of three components:
...
+```
+{
+  "nonce": 15,
+  "tool": "prediction_request",
+  "prompt": "Will my favourite football team win this week's match?"
+}
+```
+2. The application gets the metadata’s IPFS hash.
+
+3. The application writes the request’s IPFS hash to the mech contract and makes a small payment (currently $0.01 on the Gnosis chain deployment). Alternatively, the payment could be done separately through a Nevermined subscription.
```
Comment (jmoreira-valory) [inline README.md:57]:
```suggestion
3. The application writes the request’s IPFS hash to the Mech contract including a small payment (currently $0.01 on the Gnosis chain deployment). Alternatively, the payment could be done separately through a Nevermined subscription.
```

File: README.md [hunk 3]
```diff
@@ -38,6 +38,51 @@ The project consists of three components:
...
+
+4. The mech service is constantly monitoring mech contract events, and therefore gets the request hash.
```
Comment (jmoreira-valory) [inline README.md:59]:
```suggestion
4. The Mech service is constantly monitoring mech contract events, and therefore gets the request hash.
```

Comment (jmoreira-valory) [reply README.md:59]: Please change mech -> Mech

File: README.md [hunk 4]
```diff
@@ -38,6 +38,51 @@ The project consists of three components:
...
+
+5. The mech reads the request metadata from IPFS using its hash.
+
+6. The mech selects the appropriate tool to handle the request from the “tool” entry in the metadata, and runs the tool with the given arguments, usually a prompt. In this example, the mech has been requested to interact with OpenAI’s API, so it forwards the prompt to it, but the tool can implement any other desired behavior.
```
Comment (jmoreira-valory) [inline README.md:63]:
```suggestion
6. The mech selects the appropriate tool to handle the request from the `tool` entry in the metadata, and runs the tool with the given arguments, usually a prompt. In this example, the mech has been requested to interact with OpenAI’s API, so it forwards the prompt to it, but the tool can implement any other desired behavior. 
```

File: README.md [hunk 5]
```diff
@@ -38,6 +38,51 @@ The project consists of three components:
...
+
+3. The application writes the request’s IPFS hash to the mech contract and makes a small payment (currently $0.01 on the Gnosis chain deployment). Alternatively, the payment could be done separately through a Nevermined subscription.
+
+4. The mech service is constantly monitoring mech contract events, and therefore gets the request hash.
+
+5. The mech reads the request metadata from IPFS using its hash.
+
+6. The mech selects the appropriate tool to handle the request from the “tool” entry in the metadata, and runs the tool with the given arguments, usually a prompt. In this example, the mech has been requested to interact with OpenAI’s API, so it forwards the prompt to it, but the tool can implement any other desired behavior. 
+
+7. The mech gets a response from OpenAI’s API.
```
Comment (jmoreira-valory) [inline README.md:65]:
```suggestion
7. The Mech gets a response from the tool.
```

File: README.md [hunk 6]
```diff
@@ -38,6 +38,51 @@ The project consists of three components:
...
+
+8. The mech writes the response to IPFS.
```
Comment (jmoreira-valory) [inline README.md:67]:
```suggestion
8. The Mech writes the response to the IPFS.
```

File: README.md [hunk 7]
```diff
@@ -38,6 +38,51 @@ The project consists of three components:
...
+
+9. The mech receives the response IPFS hash.
+
+10. The mech writes the response hash to the mech contract.
+
+11. The application reads the response hash from the contract.
```
Comment (jmoreira-valory) [inline README.md:73]:
```suggestion
11. The application monitors for contract Deliver events and reads the response hash from the associated transaction.
```

File: README.md [hunk 8]
```diff
@@ -38,6 +38,51 @@ The project consists of three components:
...
+
+12. The application gets the response metadata from IPFS:
```
Comment (jmoreira-valory) [inline README.md:75]:
```suggestion
12. The application gets the response metadata from the IPFS:
```

File: README.md [hunk 9]
```diff
@@ -123,6 +168,31 @@ Now, you have two options to run the worker: as a standalone agent or as a servi
     bash run_service.sh
     ```
 
+## Integrating mechs into your application
+
+### For generic apps and scripts
+
+Use the [mech-client](https://github.com/valory-xyz/mech-client), that can be used either as a CLI or directly from a python script.
```
Comment (jmoreira-valory) [inline README.md:175]:
```suggestion
Use the [mech-client](https://github.com/valory-xyz/mech-client), which can be used either as a CLI or directly from a Python script.
```

File: README.md [hunk 10]
```diff
@@ -123,6 +168,31 @@ Now, you have two options to run the worker: as a standalone agent or as a servi
...
+
+### For other autonomous services
+
+To perform mech requests from your service, use the [mech_interact_abci skill](https://github.com/valory-xyz/IEKit/tree/main/packages/valory/skills/mech_interact_abci). This skill abstracts away all the IPFS and contract interactions so you only need to care about the following:
+
+-   Add the mech_interact_abci skill to your dependency list, both in packages.json, aea-config.yaml and any composed skill.yaml
```
Comment (jmoreira-valory) [inline README.md:181]:
```suggestion
-   Add the mech_interact_abci skill to your dependency list, both in `packages.json`, `aea-config.yaml` and any composed `skill.yaml`.
```

File: README.md [hunk 11]
```diff
@@ -123,6 +168,31 @@ Now, you have two options to run the worker: as a standalone agent or as a servi
...
+
+-   Import [MechInteractParams and MechResponseSpecs in your models.py file](https://github.com/valory-xyz/IEKit/blob/main/packages/valory/skills/impact_evaluator_abci/models.py#L88). You will also need to copy[ some dataclasses to your rounds.py](https://github.com/valory-xyz/IEKit/blob/main/packages/valory/skills/twitter_scoring_abci/rounds.py#L66-L97).
```
Comment (jmoreira-valory) [inline README.md:183]:
```suggestion
-   Import [MechInteractParams and MechResponseSpecs in your `models.py` file](https://github.com/valory-xyz/IEKit/blob/main/packages/valory/skills/impact_evaluator_abci/models.py#L88). You will also need to copy[ some dataclasses to your rounds.py](https://github.com/valory-xyz/IEKit/blob/main/packages/valory/skills/twitter_scoring_abci/rounds.py#L66-L97).
```

File: README.md [hunk 12]
```diff
@@ -123,6 +168,31 @@ Now, you have two options to run the worker: as a standalone agent or as a servi
...
+
+-   Add mech_requests and mech_responses to your skills' SynchonizedData class ([see here](https://github.com/valory-xyz/IEKit/blob/main/packages/valory/skills/twitter_scoring_abci/rounds.py#L181-193))
```
Comment (jmoreira-valory) [inline README.md:185]:
```suggestion
-   Add mech_requests and mech_responses to your skills' `SynchonizedData` class ([see here](https://github.com/valory-xyz/IEKit/blob/main/packages/valory/skills/twitter_scoring_abci/rounds.py#L181-193))
```

File: README.md [hunk 13]
```diff
@@ -123,6 +168,31 @@ Now, you have two options to run the worker: as a standalone agent or as a servi
...
+
+-   To send a request, [prepare the request metadata](https://github.com/valory-xyz/IEKit/blob/main/packages/valory/skills/twitter_scoring_abci/behaviours.py#L857), write it to [synchronized_data.mech_requests](https://github.com/valory-xyz/IEKit/blob/main/packages/valory/skills/twitter_scoring_abci/rounds.py#L535) and [transition into mech_interact](https://github.com/valory-xyz/IEKit/blob/main/packages/valory/skills/twitter_scoring_abci/rounds.py#L736).
```
Comment (jmoreira-valory) [inline README.md:187]:
```suggestion
-   To send a request, [prepare the request metadata](https://github.com/valory-xyz/IEKit/blob/main/packages/valory/skills/twitter_scoring_abci/behaviours.py#L857), write it to [`synchronized_data.mech_requests`](https://github.com/valory-xyz/IEKit/blob/main/packages/valory/skills/twitter_scoring_abci/rounds.py#L535) and [transition into mech_interact](https://github.com/valory-xyz/IEKit/blob/main/packages/valory/skills/twitter_scoring_abci/rounds.py#L736).
```

File: README.md [hunk 14]
```diff
@@ -123,6 +168,31 @@ Now, you have two options to run the worker: as a standalone agent or as a servi
...
+
+-   You will need to appropriately chain the mech_interact_abci skill with your other skills ([see here](https://github.com/valory-xyz/IEKit/blob/main/packages/valory/skills/impact_evaluator_abci/composition.py#L66)) and transaction_settlement_abci.
```
Comment (jmoreira-valory) [inline README.md:189]:
```suggestion
-   You will need to appropriately chain the `mech_interact_abci` skill with your other skills ([see here](https://github.com/valory-xyz/IEKit/blob/main/packages/valory/skills/impact_evaluator_abci/composition.py#L66)) and `transaction_settlement_abci`.
```

File: README.md [hunk 15]
```diff
@@ -123,6 +168,31 @@ Now, you have two options to run the worker: as a standalone agent or as a servi
...
+
+-   After the interaction finishes, the responses will be inside [synchronized_data.mech_responses](https://github.com/valory-xyz/IEKit/blob/main/packages/valory/skills/twitter_scoring_abci/behaviours.py#L903)
```
Comment (jmoreira-valory) [inline README.md:191]:
```suggestion
-   After the interaction finishes, the responses will be inside [`synchronized_data.mech_responses`](https://github.com/valory-xyz/IEKit/blob/main/packages/valory/skills/twitter_scoring_abci/behaviours.py#L903)
```

File: README.md [hunk 16]
```diff
@@ -38,6 +38,51 @@ The project consists of three components:
...
+1. Write request metadata: the application writes the request metadata to IPFS. The request can look like this:
+
+```
+{
+  "nonce": 15,
+  "tool": "prediction_request",
+  "prompt": "Will my favourite football team win this week's match?"
+}
+```
+2. The application gets the metadata’s IPFS hash.
```
Comment (jmoreira-valory) [inline README.md:55]:
```suggestion
2. The application gets the metadata's IPFS hash.
```

File: README.md (hunk 2 above)
Comment (jmoreira-valory) [inline README.md:57]:
```suggestion
3. The application writes the request's IPFS hash to the mech contract and makes a small payment (currently $0.01 on the Gnosis chain deployment). Alternatively, the payment could be done separately through a Nevermined subscription.
```

File: README.md (hunk 1 above)
Comment (adrianoVal) [reply README.md]: Updated to include the link to the image, which will become active once the image is uploaded to the folder with the merge

File: README.md [hunk 17]
```diff
@@ -38,6 +38,51 @@ The project consists of three components:
   - Gives an overview of the AI workers in the registry.
   - Allows Mech owners to create new workers.
   - Allows users to request work from an existing worker.
+ 
+## Mech request-response flow
+
+![image](https://github.com/valory-xyz/mech/blob/89572e5857dfcb69f8d4d9182df205b735448a53/docs/images/mech_request_response_flow.png)
```
Comment (jmoreira-valory) [inline README.md:44]:
```suggestion
![image](docs/images/mech_request_response_flow.png)
```

**Review by jmoreira-valory — APPROVED** [review]

**Pull request merged**